axum-extra/src/middleware/middleware_fn.rs

```diff
@@ -58,6 +58,86 @@ use tower_service::Service;
 ///     .route_layer(middleware::from_fn(auth));
 /// # let app: Router = app;
 /// ```
+///
+/// # Passing state
+///
+/// State can be passed to the function like so:
+///
+/// ```rust
+/// use axum::{
+///     Router,
+///     http::{Request, StatusCode},
+///     routing::get,
+///     response::IntoResponse,
+/// };
+/// use axum_extra::middleware::{self, Next};
+///
+/// #[derive(Clone)]
+/// struct State {
+///     in_maintenance: bool,
+/// }
+///
+/// async fn my_middleware<B>(
+///     req: Request<B>,
+///     next: Next<B>,
+///     state: State,
+/// ) -> impl IntoResponse {
+///     if state.in_maintenance {
+///         Err(StatusCode::SERVICE_UNAVAILABLE)
+///     } else {
+///         Ok(next.run(req).await)
+///     }
+/// }
+///
+/// let state = State {
```

Comment:
Wouldn't you usually want `Arc<Mutex<State>>` or something like that for this kind of state?

Reply:
Yeah you probably would. I guess removing the fields from this example well makes it less cluttered with these kinds of things. It probably depends on the exact use case exactly how you wanna share some piece of state.